File: app/src/test/java/com/coinninja/coinkeeper/service/tasks/CNHealthCheckTaskTest.java
package com.coinninja.coinkeeper.service.tasks;

import com.coinninja.coinkeeper.service.client.CoinKeeperApiClient;

import org.junit.Before;
import org.junit.Test;
import org.junit.runner.RunWith;
import org.mockito.InjectMocks;
import org.mockito.Mock;
import org.mockito.junit.MockitoJUnitRunner;

import retrofit2.Response;

import static junit.framework.Assert.assertNotNull;
import static junit.framework.Assert.assertTrue;
import static org.mockito.Mockito.verify;
import static org.mockito.Mockito.when;

@RunWith(MockitoJUnitRunner.class)
public class CNHealthCheckTaskTest {

    @Mock
    private CNHealthCheckTask.HealthCheckCallback callback;

    @Mock
    private CoinKeeperApiClient apiClient;

    @InjectMocks
    private CNHealthCheckTask task;

    @Before
    public void setUp() {
        when(apiClient.checkHealth()).thenReturn(Response.success("OK"));
    }

    @Test
    public void conducts_health_check() {
        Boolean isSuccessful = task.doInBackground();

        verify(apiClient).checkHealth();

        assertTrue(isSuccessful);
    }

    @Test
    public void successful_checks_notify_of_success() {
        task.onPostExecute(true);

        verify(callback).onHealthSuccess();
    }

    @Test
    public void failed_checks_inform_callback() {
        task.onPostExecute(false);

        verify(callback).onHealthFail();
    }

    @Test
    public void inits_instance() {
        assertNotNull(task);
        assert (task instanceof CNHealthCheckTask);
    }
}

File: app/src/test/java/com/coinninja/coinkeeper/service/tasks/CNHealthCheckTaskTest.kt
package com.coinninja.coinkeeper.service.tasks

import com.coinninja.coinkeeper.service.client.CoinKeeperApiClient
import com.google.gson.JsonObject
import com.nhaarman.mockitokotlin2.mock
import com.nhaarman.mockitokotlin2.whenever
import junit.framework.Assert.assertTrue
import org.junit.Before
import org.junit.Test
import org.mockito.Mockito.verify
import retrofit2.Response

class CNHealthCheckTaskTest {

    private val task: CNHealthCheckTask = CNHealthCheckTask(mock(), mock())

    @Before
    fun setUp() {
        whenever(task.apiClient.checkHealth()).thenReturn(Response.success(JsonObject()))
    }

    @Test
    fun conducts_health_check() {
        val isSuccessful = task.doInBackground()

        verify<CoinKeeperApiClient>(task.apiClient).checkHealth()

        assertTrue(isSuccessful!!)
    }

    @Test
    fun successful_checks_notify_of_success() {
        task.onPostExecute(true)

        verify<CNHealthCheckTask.HealthCheckCallback>(task.callback).onHealthSuccess()
    }

    @Test
    fun failed_checks_inform_callback() {
        task.onPostExecute(false)

        verify<CNHealthCheckTask.HealthCheckCallback>(task.callback).onHealthFail()
    }
}
